src/server/cluster/cluster_defs.h

```diff
@@ -24,45 +21,85 @@ struct SlotRange {
   SlotId start = 0;
   SlotId end = 0;
 
-  bool operator==(const SlotRange& r) const {
+  bool operator==(const SlotRange& r) const noexcept {
     return start == r.start && end == r.end;
   }
-  bool IsValid() {
-    return start <= end && start <= kMaxSlotId && end <= kMaxSlotId;
+
+  bool operator<(const SlotRange& r) {
+    return start < r.start || (start == r.start && end < r.end);
   }
 
-  std::string ToString() const {
-    return absl::StrCat("[", start, ", ", end, "]");
+  bool IsValid() const noexcept {
+    return start <= end && start <= kMaxSlotId && end <= kMaxSlotId;
   }
 
-  static std::string ToString(const std::vector<SlotRange>& ranges) {
-    return absl::StrJoin(ranges, ", ", [](std::string* out, SlotRange range) {
-      absl::StrAppend(out, range.ToString());
-    });
+  bool Contains(SlotId id) const noexcept {
+    return id >= start && id <= end;
   }
+
+  std::string ToString() const;
 };
 
-using SlotRanges = std::vector<SlotRange>;
+class SlotRanges {
+ public:
+  SlotRanges() = default;
+  SlotRanges(std::vector<SlotRange> ranges);
+
+  bool Contains(SlotId id) const noexcept {
+    for (const auto& sr : ranges_) {
+      if (sr.Contains(id))
+        return true;
+    }
+    return false;
+  }
+
+  size_t Size() const noexcept {
+    return ranges_.size();
+  }
+
+  bool Empty() const noexcept {
+    return ranges_.empty();
+  }
+
+  void Extend(const SlotRanges& sr);
```

Comment:
How about `Add()` or `Merge()`?

Reply:
I thought about them, but decided extend is better. If you want I can rename to Merge

Reply:
I personally prefer `Merge()` but your call, it's a nit anyway